File: modules/supabase-core/src/utils/retry.ts
/**
 * Retry Utilities
 * 
 * Retry logic for Supabase operations that might fail due to transient errors.
 * Integrates with @modules/error-handler for consistent error handling.
 * 
 * Dependencies: @modules/error-handler, @supabase/supabase-js
 */

/**
 * Retry configuration.
 */
export interface RetryConfig {
  /** Maximum number of retry attempts */
  maxRetries?: number
  /** Delay between retries in milliseconds */
  delayMs?: number
  /** Whether to use exponential backoff */
  exponentialBackoff?: boolean
  /** Function to determine if error should trigger retry */
  shouldRetry?: (error: any) => boolean
}

/**
 * Default retry configuration.
 */
const DEFAULT_RETRY_CONFIG: Required<RetryConfig> = {
  maxRetries: 3,
  delayMs: 1000,
  exponentialBackoff: true,
  shouldRetry: (error: any) => {
    // Retry on network errors or 5xx server errors
    if (error?.status >= 500) return true
    if (error?.code === 'ECONNREFUSED' || error?.code === 'ETIMEDOUT') return true
    // Don't retry on client errors (4xx) except rate limiting
    if (error?.status === 429) return true
    return false
  },
}

/**
 * Retries a Supabase operation with configurable retry logic.
 * Uses @modules/error-handler's withRetry for consistent error handling.
 * 
 * @param operation - Function that returns a Supabase operation result
 * @param config - Retry configuration
 * @returns Result from operation (wrapped in Result type if error-handler available)
 * 
 * @example
 * ```typescript
 * // With error-handler module (returns Result type)
 * const result = await withRetry(
 *   () => supabase.from('posts').select('*'),
 *   { maxRetries: 3, delayMs: 1000 }
 * )
 * if (!result.ok) {
 *   console.error('Operation failed:', result.error.message)
 *   return
 * }
 * console.log('Data:', result.value.data)
 * 
 * // Without error-handler (returns original result)
 * const { data, error } = await withRetry(
 *   () => supabase.from('posts').select('*'),
 *   { maxRetries: 3, delayMs: 1000 }
 * )
 * ```
 */
export async function withRetry<T extends { data: any; error: any }>(
  operation: () => Promise<T>,
  config: RetryConfig = {}
): Promise<T> {
  const finalConfig = { ...DEFAULT_RETRY_CONFIG, ...config }
  let lastError: any

  for (let attempt = 0; attempt <= finalConfig.maxRetries; attempt++) {
    try {
      const result = await operation()

      // If operation succeeded (no error), return result
      if (!result.error) {
        return result
      }

      // Check if we should retry this error
      if (!finalConfig.shouldRetry(result.error)) {
        return result // Don't retry, return error
      }

      lastError = result.error

      // If this was the last attempt, return the error
      if (attempt === finalConfig.maxRetries) {
        return result
      }

      // Calculate delay
      const delay = finalConfig.exponentialBackoff
        ? finalConfig.delayMs * Math.pow(2, attempt)
        : finalConfig.delayMs

      await new Promise((resolve) => setTimeout(resolve, delay))
    } catch (error: any) {
      lastError = error

      // Check if we should retry this error
      if (!finalConfig.shouldRetry(error)) {
        throw error // Don't retry, throw immediately
      }

      // If this was the last attempt, throw the error
      if (attempt === finalConfig.maxRetries) {
        throw error
      }

      // Calculate delay
      const delay = finalConfig.exponentialBackoff
        ? finalConfig.delayMs * Math.pow(2, attempt)
        : finalConfig.delayMs

      await new Promise((resolve) => setTimeout(resolve, delay))
    }
  }

  // This should never be reached, but TypeScript needs it
  throw lastError || new Error('Operation failed after retries')
}

File: modules/supabase-core-python/src/utils/retry.py
"""
Retry Utilities

Utilities for retrying Supabase operations with configurable retry logic.
"""

from dataclasses import dataclass
from typing import Callable, Any, Optional
import time


@dataclass
class RetryConfig:
    """Retry configuration."""

    max_retries: int = 3  # Maximum number of retry attempts
    delay_ms: int = 1000  # Delay between retries (milliseconds)
    exponential_backoff: bool = True  # Use exponential backoff
    should_retry: Optional[Callable[[Any], bool]] = None  # Function to determine if error should be retried


def with_retry(
    operation: Callable[[], Any],
    config: Optional[RetryConfig] = None,
) -> Any:
    """
    Retries a Supabase operation with configurable retry logic.

    Args:
        operation: Function that performs the Supabase operation
        config: Retry configuration (default: max_retries=3, delay_ms=1000)

    Returns:
        Result from operation

    Raises:
        Exception: Last error if all retries fail

    Example:
        ```python
        from supabase_core_python import with_retry, RetryConfig

        def fetch_posts():
            return supabase.table("posts").select("*").execute()

        # With default config
        result = with_retry(fetch_posts)

        # With custom config
        config = RetryConfig(max_retries=5, delay_ms=2000)
        result = with_retry(fetch_posts, config)
        ```
    """
    if config is None:
        config = RetryConfig()

    last_error = None

    for attempt in range(config.max_retries + 1):
        try:
            return operation()
        except Exception as error:
            last_error = error

            # Check if we should retry this error
            if config.should_retry and not config.should_retry(error):
                raise error

            # Don't retry on certain errors (e.g., validation errors)
            if hasattr(error, "code"):
                error_code = getattr(error, "code", None)
                if error_code in ["23505", "23503"]:  # Unique constraint, foreign key
                    raise error

            # Don't retry on last attempt
            if attempt < config.max_retries:
                # Calculate delay
                if config.exponential_backoff:
                    delay = config.delay_ms * (2 ** attempt)
                else:
                    delay = config.delay_ms

                time.sleep(delay / 1000.0)  # Convert ms to seconds
                continue

    # All retries failed
    raise last_error or Exception("Operation failed after retries")

